include/revng/ABI/CallingConventionTrait.h

```diff
@@ -0,0 +1,1083 @@
+  static constexpr size_t MaximumGPRsPerAggregateReturnValue = 0;
+  static constexpr size_t MaximumGPRsPerScalarArgument = 0;
+  static constexpr size_t MaximumGPRsPerScalarReturnValue = 0;
+
+  static constexpr size_t MinimumStackArgumentSize = 8;
+
+  static constexpr std::array<model::Register::Values, 0>
+    GeneralPurposeArgumentRegisters = {};
+  static constexpr std::array<model::Register::Values, 0>
+    GeneralPurposeReturnValueRegisters = {};
+
+  static constexpr std::array<model::Register::Values, 0>
+    VectorArgumentRegisters = {};
+  static constexpr std::array<model::Register::Values, 0>
+    VectorReturnValueRegisters = {};
+
+  static constexpr std::array CalleeSavedRegisters = {
+    model::Register::r12_x86_64,  model::Register::r13_x86_64,
+    model::Register::r14_x86_64,  model::Register::r15_x86_64,
+    model::Register::rdi_x86_64,  model::Register::rsi_x86_64,
+    model::Register::rbx_x86_64,  model::Register::rbp_x86_64,
+    model::Register::xmm6_x86_64, model::Register::xmm7_x86_64
+  };
+
+  static constexpr auto ReturnValueLocationRegister = model::Register::Invalid;
+};
+
+/// The default SystemV x86 ABI
+/// The parameters are only passed using the stack.
+/// \todo: look into the st-return values.
+template<>
+struct CallingConventionTrait<model::abi::SystemV_x86> {
+  static constexpr bool ArgumentsArePositionBased = false;
+  static constexpr bool OnlyStartDoubleArgumentsFromAnEvenRegister = false;
+  static constexpr bool ArgumentsCanBeSplitBetweenRegistersAndStack = false;
+  static constexpr bool UsePointerToCopyForStackArguments = false;
+
+  static constexpr size_t MaximumGPRsPerAggregateArgument = 0;
+  static constexpr size_t MaximumGPRsPerAggregateReturnValue = 0;
+  static constexpr size_t MaximumGPRsPerScalarArgument = 0;
+  static constexpr size_t MaximumGPRsPerScalarReturnValue = 2;
+
+  static constexpr size_t MinimumStackArgumentSize = 4;
+
+  static constexpr std::array<model::Register::Values, 0>
+    GeneralPurposeArgumentRegisters = {};
+  static constexpr std::array GeneralPurposeReturnValueRegisters = {
+    model::Register::eax_x86,
+    model::Register::edx_x86
+  };
+
+  static constexpr std::array VectorArgumentRegisters = {
+    model::Register::xmm0_x86,
+    model::Register::xmm1_x86,
+    model::Register::xmm2_x86
+  };
+  static constexpr std::array VectorReturnValueRegisters = {
+    // model::Register::st0_x86,
+    model::Register::xmm0_x86
+  };
+
+  static constexpr std::array CalleeSavedRegisters = {
+    model::Register::ebx_x86,
+    model::Register::ebp_x86,
+    model::Register::esp_x86,
+    model::Register::edi_x86,
+    model::Register::esi_x86
+  };
+
+  static constexpr auto ReturnValueLocationRegister = model::Register::eax_x86;
+};
+
+/// `regparm(1)` SystemV x86 ABI
+/// Function arguments are passed using both stack and specified registers.
+template<>
+struct CallingConventionTrait<model::abi::SystemV_x86_regparm_1> {
```

Comment:
Is the `regparm(1)` a loosely defined compiler extension or is it documented in an """official""" ABI document?

Reply:
It's definitely a loose `gcc` extension. I just thought that adding it doesn't really require any additional effort and it doesn't really harm anyone.

Reply:
And by loose it's meant, that's it's pretty much the normal `SystemV_x86` ABI but with extra register parameters strapped on top.

Reply:
Gotcha, interesting.